rs/boundary_node/rate_limits/canister/fetcher.rs

```diff
@@ -0,0 +1,190 @@
+use crate::{
+    confidentiality_formatting::ConfidentialityFormatting,
+    state::Repository,
+    types::{ConfigResponse, OutputConfig, OutputRule, OutputRuleMetadata, RuleId, Version},
+};
+
+pub trait EntityFetcher {
+    type Input;
+    type Output;
+    type Error;
+
+    fn fetch(&self, input: Self::Input) -> Result<Self::Output, Self::Error>;
+}
+
+pub struct ConfigFetcher<R, F> {
+    pub repository: R,
+    pub formatter: F,
+}
+
+pub struct RuleFetcher<R, F> {
+    pub repository: R,
+    pub formatter: F,
+}
+
+impl<R, F> RuleFetcher<R, F> {
+    pub fn new(repository: R, formatter: F) -> Self {
+        Self {
+            repository,
+            formatter,
+        }
+    }
+}
+
+impl<R, F> ConfigFetcher<R, F> {
+    pub fn new(repository: R, formatter: F) -> Self {
+        Self {
+            repository,
+            formatter,
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum FetchConfigError {
+    #[error("Config for version={0} not found")]
+    NotFound(Version),
+    #[error("No existing config versions")]
+    NoExistingVersions,
+    #[error(transparent)]
+    Unexpected(#[from] anyhow::Error),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum FetchRuleError {
+    #[error("Rule with id={0} not found")]
+    NotFound(RuleId),
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+impl<R: Repository, F: ConfidentialityFormatting<Input = OutputConfig>> EntityFetcher
+    for ConfigFetcher<R, F>
+{
+    type Input = Option<Version>;
+    type Output = ConfigResponse;
+    type Error = FetchConfigError;
+
+    fn fetch(&self, version: Option<Version>) -> Result<ConfigResponse, FetchConfigError> {
+        let current_version = self
+            .repository
+            .get_version()
+            .ok_or_else(|| FetchConfigError::NoExistingVersions)?;
+
+        let version = version.unwrap_or(current_version.0);
+
+        let stored_config = self
+            .repository
+            .get_config(version)
+            .ok_or_else(|| FetchConfigError::NotFound(version))?;
+
+        let mut rules: Vec<OutputRule> = vec![];
+
+        for rule_id in stored_config.rule_ids.iter() {
+            let rule = self.repository.get_rule(rule_id).ok_or_else(|| {
+                FetchConfigError::Unexpected(anyhow::anyhow!("Rule with id = {rule_id} not found"))
+            })?;
+
+            let output_rule = OutputRule {
+                id: rule_id.clone(),
+                incident_id: rule.incident_id,
+                rule_raw: Some(rule.rule_raw),
+                description: Some(rule.description),
+                disclosed_at: rule.disclosed_at,
+            };
+
+            rules.push(output_rule);
+        }
+
+        let config = OutputConfig {
+            schema_version: stored_config.schema_version,
+            rules,
+        };
+
+        let formatted_config = self.formatter.format(&config);
+
+        let config = ConfigResponse {
```

Comment:
Maybe this config should be renamed such that we never return the OutputConfig from above directly.